solr/core/src/java/org/apache/solr/handler/component/TimeLimitingHttpShardHandler.java

```diff
@@ -0,0 +1,328 @@
+    this.timeoutCallback = timeoutCallback;
+  }
+
+  @Override
+  public void onRequestSubmitted(List<String> shardUrls, Future<?> future) {
+    pendingFutures.add(future);
+
+    // This is a shardRequest being submitted to a single shard. However with multiple replicas,
+    // there will be multiple shard URLs.
+    // if any of those replicas are from a known slow node, then we will NOT count this pending
+    // future as a "fast node future". Therefore, we assume this future can be potentially slow
+    // and subjected to timeout.
+    if (!hasSlowNodeShardUrl(shardUrls)) {
+      pendingFutureCountFromFastNode.incrementAndGet();
+    }
+  }
+
+  private boolean hasSlowNodeShardUrl(List<String> shardUrls) {
+    for (String shardUrl : shardUrls) {
+      if (slowNodes.contains(Util.getNode(shardUrl))) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  @Override
+  public synchronized boolean onRequestCompleted(
+      String selectedShardUrl, List<String> shardUrls, Future<?> future, long timeElapsed) {
+    pendingFutures.remove(future);
+    if (pendingFutures.isEmpty()) { // every submitted futures are processed
+      if (timeoutTask != null) {
+        timeoutTask.cancel(true);
+      }
+      return true;
+    }
+
+    if (!hasSlowNodeShardUrl(shardUrls)) {
+      pendingFutureCountFromFastNode.decrementAndGet();
+    }
+
+    if (timeoutTask == null
+        && pendingFutureCountFromFastNode.get()
+            <= 0) { // all pending reqs are from slow nodes, start a timer to possibly timeout
+      ExecutorService executorService = null;
+      try {
+        executorService =
+            ExecutorUtil.newMDCAwareSingleThreadExecutor(
+                new SolrNamedThreadFactory("SlowNodeTimeout"));
+        timeoutTask =
+            executorService.submit(
+                () -> {
+                  try {
+                    Thread.sleep(timeout);
+                  } catch (InterruptedException e) {
+                    // ok. The responses came back before timeout
+                    return;
+                  }
+                  Set<Future<?>> removingFutures = new HashSet<>(pendingFutures);
+                  pendingFutures.clear();
+                  if (!removingFutures.isEmpty()) {
+                    if (timeoutCallback != null) {
+                      timeoutCallback.accept(removingFutures);
+                    }
+                    if (!dryRun) {
+                      if (log.isInfoEnabled()) {
+                        log.info(
+                            "Cancelling {} pending requests due to timeout duration {}ms exceeded. ",
+                            removingFutures.size(),
+                            timeout);
+                      }
+                      removingFutures.forEach(f -> f.cancel(true));
+                      if (log.isInfoEnabled()) {
+                        log.info("{} Pending requests cancelled", removingFutures.size());
+                      }
```

Comment:
do we really need to log at INFO level before _and_ after, essentially the same message?

Reply:
Good point! Consolidate it to a single message after they are cancelled